typescript/sdk/src/ism/EvmIsmModule.ts

```diff
@@ -1,89 +1,727 @@
   protected constructor(
     protected readonly multiProvider: MultiProvider,
     protected readonly deployer: HyperlaneDeployer<any, any>,
     args: HyperlaneModuleArgs<
       IsmConfig,
-      HyperlaneContracts<ProxyFactoryFactories> & {
-        deployedIsm: Address;
-      }
+      HyperlaneAddresses<ProxyFactoryFactories> & ExtraArgs
     >,
   ) {
     super(args);
+
     this.reader = new EvmIsmReader(multiProvider, args.chain);
-    this.creator = new EvmIsmCreator(deployer, multiProvider, args.addresses);
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { mailbox, deployedIsm, ...addresses } = args.addresses;
+    this.factories = attachAndConnectContracts(
+      addresses,
+      proxyFactoryFactories,
+      multiProvider.getSigner(args.chain),
+    );
+
+    this.chainName = this.multiProvider.getChainName(this.args.chain);
+    this.domainId = this.multiProvider.getDomainId(this.chainName);
   }
 
   public async read(): Promise<IsmConfig> {
-    return await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+    return this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
   }
 
-  public async update(config: IsmConfig): Promise<EthersV5Transaction[]> {
-    throw new Error('Method not implemented.');
+  // whoever calls update() needs to ensure that targetConfig has a valid owner
+  public async update(
+    targetConfig: IsmConfig,
+  ): Promise<AnnotatedEthersV5Transaction[]> {
+    const currentConfig =
+      typeof this.args.config === 'string'
+        ? this.args.addresses.deployedIsm
+        : await this.read();
 
-    const destination = this.multiProvider.getChainName(this.args.chain);
-    await this.creator.update({
-      destination,
-      config,
-      existingIsmAddress: this.args.addresses.deployedIsm,
+    // Update the config in case it's a custom ISM
+    this.args.config = targetConfig;
+
+    // moduleMatchesConfig expects any domain filtering to have been done already
+    if (
+      typeof targetConfig !== 'string' &&
+      (targetConfig.type === IsmType.ROUTING ||
+        targetConfig.type === IsmType.FALLBACK_ROUTING)
+    ) {
+      // filter for known domains
+      const { availableDomains } = this.filterRoutingIsmDomains({
+        config: targetConfig,
+      });
+      targetConfig.domains = availableDomains;
+    }
+
+    // this calls resolveOrDeployAccountOwner under the hood
+    const configMatches = await moduleMatchesConfig(
+      this.chainName,
+      this.args.addresses.deployedIsm,
+      targetConfig,
+      this.multiProvider,
+      this.factories,
+      this.args.addresses.mailbox,
+    );
+
+    // If configs match, no updates needed
+    if (configMatches) {
+      return [];
+    }
+
+    // Else, we have to figure out what an update for this ISM entails
+
+    // If target config is a custom ISM, just update the address
+    // if config -> custom ISM, update address
+    // if custom ISM -> custom ISM, update address
```

Comment:
I dont understand these comments

Reply:
custom ISM is when the given config is just a string

if we're going from a string -> config, need to a full deploy

if we're going config -> config, do the update logic

if we're going config -> string or string -> string, no need to do any further logic besides updating the address of the ism in state

Reply:
can you replace "custom ISM" with "address ISM" then?